modules/ensemble/lib/action/Log_event_action.dart

```diff
@@ -7,31 +7,47 @@ import 'package:ensemble/util/utils.dart';
 import 'package:ensemble_ts_interpreter/invokables/invokable.dart';
 import 'package:flutter/cupertino.dart';
 
-class LogEvent extends EnsembleAction {
-  final String eventName;
+class TrackEvent extends EnsembleAction {
+  final String? eventName;
   final Map<dynamic, dynamic>? parameters;
+  final String? provider;
+  final String? operation;
+  final String? userId;
   final String logLevel;
-  LogEvent(
-      {super.initiator,
-      required this.eventName,
-      required this.logLevel,
-      this.parameters});
 
-  factory LogEvent.from({Invokable? initiator, Map? payload}) {
+  TrackEvent({
+    required Invokable? initiator,
+    this.eventName,
+    required this.logLevel,
+    required this.provider,
+    this.operation,
+    this.userId,
+    this.parameters,
+  }) : super(initiator: initiator);
+
+  factory TrackEvent.from({Invokable? initiator, Map? payload}) {
     payload = Utils.convertYamlToDart(payload);
     String? eventName = payload?['name'];
-    if (eventName == null) {
+    String? operation = payload?['operation'];
+
+    if (operation == 'logEvent' && eventName == null) {
       throw LanguageError(
-          "${ActionType.logEvent.name} requires the event name");
+          "${ActionType.trackEvent.name} requires the event name");
+    } else if (operation == 'setUserId' && payload?['userId'] == null) {
```

Comment:
this is not object oriented programming. Don't overload the same logevent by adding operation and then changing all the code everywhere to handle it. Have a base class that's LogEvent and extend it with LogSetUserIdEvent and set whatever you need to set in that. 

Also have a clear logSetUserId as a method on the provider so you don't have to hide things in parameters. Fix this whole thing up. better to start fresh

Reply:
setUserId is supposed to be called only once when user starts his session i.e after login or when user reopens the app, not with every event, that's why I added it as operation. + we might have to support more operations as well, as client was also requesting for `setUserProperty` which they realized its not needed right now, but later on they would need it.

Firebase Analytics has a lot of operations, and it the current structure is not flexible enough to add more in future. We have to add operations to make it easier for app dev side.